pkg/codegen/schema/bind.go

```diff
@@ -548,6 +548,11 @@ func (spec *PackageSpec) validateTypeTokens() hcl.Diagnostics {
 	for t := range spec.Types {
 		diags = diags.Extend(spec.validateTypeToken(allowedPackageNames, "types", t))
 	}
+
+	// When validating function type tokens, we'll add `pulumi` to the list of allowed package names in order to support
+	// defining methods *on provider resources themselves*. In these cases, we'll see tokens of the form
+	// `pulumi:providers:<package>/<method>`, which we want to treat as valid.
+	allowedPackageNames["pulumi"] = true
```

Comment:
This is ok, but we should probably be stricter here that the tokens are exactly pulumi:providers:pkg. Maybe issue that so we do fix it when we've got a day for it.

Reply:
I think I've managed to make this work by tweaking the map of allowed package names so that rather than just mapping to booleans and tracking membership, it now maps to allowed module names (where `nil` signals "all module names"). In doing so I think we can strengthen this check as part of this work rather than in a follow-up. I've extended the validation unit tests and I think it's covered (as well as a missing case for reserved provider identifiers). Let me know what you think!

Reply:
(https://github.com/pulumi/pulumi/pull/18920/files#diff-9ab9df002b36270fe280ee8e8d162d61a0946708e23a60733f30159b8727b937R516)